runtime/src/main/java/org/corfudb/runtime/LogReplicationListener.java

```diff
@@ -0,0 +1,184 @@
+    }
+
+    public LogReplicationListener(CorfuStore corfuStore, @Nonnull String namespace, @Nonnull String streamTag,
+                                  List<String> tablesOfInterest, int bufferSize) {
+        this.corfuStore = corfuStore;
+        this.namespace = namespace;
+        this.streamTag = streamTag;
+        this.tablesOfInterest = tablesOfInterest;
+        this.bufferSize = bufferSize;
+    }
+
+    /**
+     * This is an internal method of this abstract listener and not exposed to clients.
+     *
+     * @param results is a map of stream UUID -> list of entries of this stream.
+     */
+    public final void onNext(CorfuStreamEntries results) {
+
+        // A corfu update can/may have multiple updates belonging to different streams.  This callback will return
+        // those updates as a list grouped by their Stream UUIDs.
+        // Note: there is no order guarantee within the transaction boundaries.
+
+        Set<String> tableNames =
+                results.getEntries().keySet().stream().map(schema -> schema.getTableName()).collect(Collectors.toSet());
+
+        if (tableNames.contains(REPLICATION_STATUS_TABLE)) {
+            Preconditions.checkState(results.getEntries().keySet().size() == 1,
+                "Replication Status Table Update received with other tables");
+            processReplicationStatusUpdate(results);
+            return;
+        }
+
+        // Data Updates
+        if (snapshotSyncInProgressOnSubscription.get()) {
+            // If the listener started when snapshot sync was already ongoing, ignore all data updates until it ends
+            return;
+        }
+
+        if (snapshotSyncInProgress.get()) {
+            processUpdatesInSnapshotSync(results);
+        } else {
+            processUpdatesInLogEntrySync(results);
+        }
+    }
+
+    private void processReplicationStatusUpdate(CorfuStreamEntries results) {
+        Map<TableSchema, List<CorfuStreamEntry>> entries = results.getEntries();
+
+        TableSchema replicationStatusTableSchema =
+                entries.keySet().stream().filter(key -> key.getTableName().equals(REPLICATION_STATUS_TABLE))
+                        .findFirst().get();
+
+        for (CorfuStreamEntry entry : entries.get(replicationStatusTableSchema)) {
+
+            // Ignore any update where the operation type != UPDATE
+            if (entry.getOperation() == CorfuStreamEntry.OperationType.UPDATE) {
+                ReplicationStatusVal status = (ReplicationStatusVal)entry.getPayload();
+
+                if (snapshotSyncInProgressOnSubscription.get() && status.getDataConsistent() == true) {
+                    snapshotSyncInProgressOnSubscription.set(false);
+                    corfuStore.unsubscribeListener(this);
+
+                    if (bufferSize == 0) {
+                        corfuStore.subscribeLogReplicationListener(this, namespace, streamTag,
+                                tablesOfInterest);
```

Comment:
Still tablesOfInterest could be null.

Reply:
Would it be cleaner to use the constructor with all params and internally handle empty lists, <=0 buffer size ...?